### PR TITLE
libstore: cache AWS credentials and refresh on S3 token expiry

### PR DESCRIPTION
**Pull request by lovesegfault**
Follow-up to #15594. Wraps the credential provider chain in `CredentialsProviderCached` (15-minute TTL, honors embedded expiration if shorter). Previously every S3 request triggered a fresh STS / SSO / IMDS round-trip — for a build pulling hundreds of narinfos that was hundreds of STS calls. aws-c-auth's default chain does this internally; our custom chain bypassed it.

When an S3 request fails with an `ExpiredToken` or `TokenRefreshRequired` error code, credentials are refreshed asynchronously and the request retried. Previously a session token expiring mid-build was a hard failure.

**Trigger:**
- `finish()` inspects the S3 error body captured in `errorSink` for `<Code>ExpiredToken</Code>` or `<Code>TokenRefreshRequired</Code>`. AWS S3 [documents these as HTTP 400](https://docs.aws.amazon.com/AmazonS3/latest/API/ErrorResponses.html), so the check admits both 400 and 403 — AWS SDKs check the body `<Code>` status-agnostic for this reason. `AccessDenied` (403, missing key in an unlistable bucket) does **not** match — those probes stay at one S3 request, preserving the existing 403→404 mapping at `http-binary-cache-store.cc:173,280,302`.
- HEAD requests have no body to inspect and fall through — not a regression, HEAD never retried before either.
- 401 and 407 are always `Forbidden` — S3 never sends 401, and 407 is proxy auth which a credential refresh cannot fix.

**Cold-start:**
- `setupForS3()` tries `tryGetCredentials` once synchronously (hot cache is instant). It always sets `awsRetry` in the provider branch, even if the synchronous fetch failed; in that case `usernameAuth` stays empty and the first unauthenticated 403 (`AccessDenied`) triggers the same async refresh path — `s3TokenExpired()` returns true on `!usernameAuth` before the body check.
- Bounded by `awsCredsRefreshed`: `init()` sets it unconditionally when the future resolves (even nullopt), and even a blackholed-IMDS chain resolves within ~1s — well inside the retry window. `download-attempts` is the backstop for a genuinely-hung provider.
- One `warnOnce` at the terminal `!canRetry` site if `usernameAuth` is still empty — tells the user "provider" not "permissions". Not in the pre-resolved sandbox path (`builtin:fetchurl` — `awsRetry` is never set there).

**Refresh mechanism:**
- `maybeRetry()` spawns a detached `std::thread` + `std::promise`. A `std::promise`-sourced future does not block in its destructor (unlike `std::async`, whose future joins the launched thread — [futures.async]/5).
- The detached thread calls `tryGetCredentials` (noexcept, no `warn()`). `nix::logger` may be destroyed before this thread completes since it isn't joined by any destructor — the AWS CRT's process-global log hook is a residual risk at `-vvvvv`, but default verbosity is gated.
- The provider is stashed on `FileTransferRequest` as `AwsRetryInfo {ParsedS3URL, ref<AwsCredentialProvider>}`, set by `setupForS3()` on the caller thread, then init-captured by value into the detached thread — `getAwsCredentialsProvider()`'s function-local static is constructed after the `fileTransfer` static and destroyed before `~curlFileTransfer()` joins the worker (LIFO), so neither the worker thread nor the detached thread touch it.
- `init()` applies the result via `wait_for(0)` — non-blocking. The retry loop's `computeRetryDelayMs` backoff overlaps with the fetch.
- The spawn is guarded against `std::system_error` (e.g. `EAGAIN` under `ulimit -u`); `get_future()` is assigned after `.detach()` succeeds so a constructor throw doesn't leave a dangling future.

**Termination:**
- `awsCredsRefreshed` is set in `init()` when the future resolves; a consecutive expiry after that falls through (the refresh didn't help — clock skew, server-side revocation).
- The flag is cleared **only on 2xx** in `headerCallback` — not 1xx (`100 Continue` on PUT uploads isn't auth confirmation; some S3-compatibles send it before validating the header), not 3xx (a redirect hop via `CURLOPT_FOLLOWLOCATION` doesn't validate auth either). So a long Range-resumable download spanning multiple token lifetimes can refresh more than once.
- If retries exhaust before the refresh completes, `exc.error` is downgraded — 403→`Forbidden` (unlistable-bucket mapping), 400→`Misc` (unambiguous credential failure → surface it). Only when `writtenToSink == 0`: a partial 200 OK proves the file exists, so `Forbidden`→`NoSuchBinaryCacheFile` would lie.

**Range-resume / proxy:**
- `acceptRanges` is preserved across 1xx/3xx/4xx/5xx (only 2xx resets, same gate as the loop-guard). An error response doesn't carry `Accept-Ranges`, but it also doesn't revoke the server's range support established by a prior 200; nor does a redirect hop or a Continue.
- `CURLOPT_SUPPRESS_CONNECT_HEADERS` so a proxy's "200 Connection established" doesn't reach `headerCallback` and clobber the same per-response state.

**Known limitation:**
- The cached provider honors embedded credential expiration, so natural expiry works. Under clock skew or server-side token revocation, the refresh may return the same cached credential — we don't force-invalidate since the `<Code>` gate already bounds this path to cases where the provider's TTL should align with S3's view.

### CODE REVIEW

File: src/libstore/filetransfer.cc [hunk 1]
```diff
@@ -758,10 +829,42 @@ struct curlFileTransfer : public FileTransfer
...
+                                 return httpStatus == HttpStatus::Forbidden;
+                             if (httpStatus != HttpStatus::Forbidden && httpStatus != HttpStatus::BadRequest)
+                                 return false;
+                             if (!errorSink)
+                                 return false;
+                             static const std::regex codeRe(
+                                 "<Code>(ExpiredToken|TokenRefreshRequired)</Code>", std::regex::optimize);
+                             return std::regex_search(errorSink->s, codeRe);
+                         }()) {
+                    err = Transient;
```
Comment (xokdvium) [inline src/libstore/filetransfer.cc]: Can this become more hookable? Not sure want the-specific logic to be directly in the client loop. Maybe an optional callback would be a nicer fit here.

File: src/libstore/filetransfer.cc [hunk 2]
```diff
@@ -875,10 +978,64 @@ struct curlFileTransfer : public FileTransfer
...
+            // direct logger deref from warn(), but the AWS CRT log hook
+            // (nixAwsLoggerLog) can still touch nix::logger at high verbosity
+            // during a slow fetch — residual risk at -vvvvv, not at default.
+            if ((httpStatus == HttpStatus::Forbidden || httpStatus == HttpStatus::BadRequest) && request.awsRetry
+                && !pendingAwsCredRefresh) {
+                try {
+                    auto promise = std::make_shared<std::promise<std::optional<AwsCredentials>>>();
+                    std::thread([url = request.awsRetry->s3Url, promise, provider = request.awsRetry->provider]() {
+                        promise->set_value(provider->tryGetCredentials(url));
+                    }).detach();
```
Comment (xokdvium) [inline src/libstore/filetransfer.cc]: Detaching has a big risk of static destruction order fiasco. We've already seen this be a huge issue with autoGC.

File: src/libstore/filetransfer.cc [hunk 3]
```diff
@@ -565,6 +610,28 @@ struct curlFileTransfer : public FileTransfer
 
             curl_easy_reset(req);
 
+#if NIX_WITH_AWS_AUTH
+            if (pendingAwsCredRefresh
+                && pendingAwsCredRefresh->wait_for(std::chrono::seconds(0)) == std::future_status::ready) {
+                auto creds = pendingAwsCredRefresh->get();
+                pendingAwsCredRefresh.reset();
+                awsCredsRefreshed = true;
```
Comment (xokdvium) [inline src/libstore/filetransfer.cc]: What we can do here is do something similar to the pausing mechanism and completely avoid working the requests that are awaiting some work. Something like the cross-thread wake-up queue that the pausing wake-up is using.

**Comment by cep21** [comment]
Hi,

I ran into the same issue here.  Any updates?
